src/main/java/com/datatorrent/netlet/OptimizedEventLoop.java

```diff
@@ -0,0 +1,194 @@
+      if (pos >= keys.length) {
+        SelectionKey[] keys = new SelectionKey[this.keys.length << 1];
+        System.arraycopy(this.keys, 0, keys, 0, this.keys.length);
+        this.keys = keys;
+      }
+      keys[pos++] = key;
+      return true;
+    }
+
+    @Override
+    public int size()
+    {
+      return pos;
+    }
+
+    public boolean remove(Object o) {
+      return false;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+      if (o == null) {
+        return false;
+      }
+      for (int i = 0; i < pos; i++) {
+        if (o.equals(keys[i])) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    @Override
+    public Iterator<SelectionKey> iterator() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  public OptimizedEventLoop(String id) throws IOException
+  {
+    super(id);
+    try {
+      ClassLoader systemClassLoader;
+      if (System.getSecurityManager() == null) {
+        systemClassLoader = ClassLoader.getSystemClassLoader();
+      } else {
+        systemClassLoader = AccessController.doPrivileged(new PrivilegedAction<ClassLoader>()
+        {
+          @Override
+          public ClassLoader run()
+          {
+            return ClassLoader.getSystemClassLoader();
+          }
+        });
+      }
+
+      final Class<?> selectorClass = Class.forName("sun.nio.ch.SelectorImpl", false, systemClassLoader);
+      if (selectorClass.isAssignableFrom(selector.getClass())) {
+        Field selectedKeys = selectorClass.getDeclaredField("selectedKeys");
+        Field publicSelectedKeys = selectorClass.getDeclaredField("publicSelectedKeys");
+        selectedKeys.setAccessible(true);
+        publicSelectedKeys.setAccessible(true);
+        SelectedSelectionKeySet keys = new SelectedSelectionKeySet(1024);
+        selectedKeys.set(selector, keys);
+        publicSelectedKeys.set(selector, keys);
+        logger.debug("Instrumented an optimized java.util.Set into: {}", selector);
+      }
+    }
+    catch (Exception e) {
+      logger.debug("Failed to instrument an optimized java.util.Set into: {}", selector, e);
+    }
+  }
+
+  @SuppressWarnings({"SleepWhileInLoop", "ConstantConditions"})
+  protected void runEventLoop()
+  {
+    Set<SelectionKey> selectedKeys = selector.selectedKeys();
+    if (selectedKeys instanceof SelectedSelectionKeySet) {
+      runEventLoop((SelectedSelectionKeySet) selectedKeys);
+    } else
+      super.runEventLoop();
+  }
+
+  private void runEventLoop(SelectedSelectionKeySet keys)
+  {
+    do {
+      SelectionKey sk = null;
+      try {
+        do {
+          int size = tasks.size();
+          while (alive && size > 0 && selector.selectNow() == 0) {
+            tasks.pollUnsafe().run();
+            size--;
+          }
+          if (alive && size == 0 && selector.select() == 0) {
+            continue;
+          }
+
+          SelectionKey[] selectedKeys = keys.getKeys();
+          for (int i = 0; alive; i++) {
```

Comment:
I think that it will be okay for the all the selectors selected in the current cycle are processed. That should also give the thread some processing boost as it avoids access to volatile variable.